### PR TITLE
Several Important Changes

### PR DESCRIPTION
**Pull request by digital-flowers**
1- Change C++ GetContacts to Async
i have noticed that the wire electron application get totally freeze when you pull the user contacts specially if the user has many contacts in my case (500), so i rewrite the node add-on getContacts cpp method to use Async Progressive Worker so the new getContacts method signature is getContacts(onProgress, onFinish) and it working smooth now

2- Use Nan for cleaner code and more backward compatibility

3- Write javascript interface for easier use specially for IDE's

### CODE REVIEW

File: example.js [hunk 1]
```diff
@@ -1,57 +1,24 @@
...
-// (at your option) any later version.
-//
-// This program is distributed in the hope that it will be useful,
-// but WITHOUT ANY WARRANTY; without even the implied warranty of
-// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
-// GNU General Public License for more details.
-//
-// You should have received a copy of the GNU General Public License
-// along with this program. If not, see <http://www.gnu.org/licenses/>.
+const readline = require('readline');
```
Comment (marcoconti83) [inline example.js]: Please do not remove the copyright header

Comment (digital-flowers) [inline example.js]: ok do you want me to fix this or you will do it

File: wrapper.cc [hunk 2]
```diff
@@ -15,119 +15,133 @@
...
-	Person *p = ab.getContact(i);
-	fillPersonObject(isolate, contact, p);
-	delete p;
-
-	args.GetReturnValue().Set(contact);
+    AddressBook ab;
+    Isolate* isolate = Isolate::GetCurrent();
+
+    Local<Object> contact = Object::New(isolate);
+    fillPersonObject(isolate, contact, ab.getContact(index));
```
Comment (c-g-owen) [inline wrapper.cc]: This code crashes if an invalid index is given, which happens in example.js if access to contacts isn't granted. Should either check `index => 0 && index < ab.contactCount()` and/or that `ab.getContact()` doesn't return NULL.

File: README.md [hunk 3]
```diff
@@ -19,5 +19,20 @@ Building for electron seems to require the use of electron-rebuild:
 
 # Usage
 
-See example.js for how to use the module
+how to use the module:
```
Comment (mikeger) [inline README.md]: I would stick to document style with capitalization, i.e. "How to use the module:"

File: example.js [hunk 4]
```diff
@@ -1,57 +1,24 @@
...
+console.log("Me: ".custom, addressBook.getMe());
 
-// Addressbook has a member me that represents the current user.
-// contacts member is an array of all contacts (includes the me user)
-//
-// me and contact objects have the following members:
-// firstName - string
-// lastName - string
-// emails - array of strings of email addresses
-// numbers - array of strings of phone numbers
```
Comment (mikeger) [inline example.js]: Did the fields remain the same? In this case I would leave this comment, seems useful.

File: example.js (hunk 1 above)
Comment (marcoconti83) [inline example.js]: It would be great if you could re-add the deleted lines

File: wrapper.cc [hunk 5]
```diff
@@ -15,119 +15,133 @@
...
-		Local<String> result = String::NewFromUtf8(isolate, src[i].c_str());
-		array->Set(i, result);
-	}	
-	obj->Set(String::NewFromUtf8(isolate, name), array);
+    Local<Array> array = Array::New(isolate);
+    for (unsigned int i = 0; i < src.size(); i++ ) {
+        Local<String> result = String::NewFromUtf8(isolate, src[i].c_str());
+        array->Set(i, result);
+    }
+    obj->Set(String::NewFromUtf8(isolate, name), array);
```
Comment (lipis) [inline wrapper.cc]: I created the #4 to avoid those..

File: example.js (hunk 1 above)
Comment (digital-flowers) [inline example.js]: yes sure

File: example.js (hunk 4 above)
Comment (digital-flowers) [inline example.js]: yes all the fields is same as before

Comment (mikeger) [inline example.js]: Cool, let's have the readme then 👍

File: wrapper.cc (hunk 5 above)
Comment (digital-flowers) [inline wrapper.cc]: nice i have learned something new from this editor config thing 👍

**Comment by digital-flowers** [comment]
what about now ?

**Comment by marcoconti83** [comment]
Thanks for re-adding the headers! 👍

**Comment by herrmannplatz** [comment]
@lipis Would You Mind to create a desktop build with this branch so that QA can have a look

**Comment by lipis** [comment]
Actually, it didn't work on my machine :( Was throwing an error.. did you try to include it in the Desktop client @digital-flowers? (I'll prepare a build anyway.. maybe it's something with my rights)

**Comment by lipis** [comment]
Here is the error message.. screen shot 2016-12-22 at 11 38 35

**Comment by digital-flowers** [comment]
sorry as i mentioned in my commit after my changes address book javascript interface changed 
window.zAddressBook is not a function it is an object with the following methods: 
getContact(index);
getMe();
getContactsCount();
getContacts(onProgress, onFinish);

the previous addressbook was a function that pull all the contacts synchronously  once you include it even if you want to access one contact and this is why i changed it to an object
so now you need to to do 
```javascript
window.zAddressBook.getContacts(function(percentage){
    // here you can provide visual feedback
    console.log(percentage+"%");
}, function(contacts){
    // your contacts is here
    console.log(contacts);
})
```

**Comment by digital-flowers** [comment]
i already using it inside electron and it is working perfectly !

**Comment by lipis** [comment]
Awesome. Stupid me.. I'll create a new PR and we'll take it ftom there.. thanks a bunch..

**Comment by lipis** [comment]
Sorry for the long delay.. but we finally tested it! Works like a charm.. 

- https://github.com/wireapp/wire-webapp/pull/648
- https://github.com/wireapp/wire-desktop/pull/332

File: .gitignore [hunk 6]
```diff
@@ -1,7 +1,8 @@
...
 .*.swp
+*.log.*
 local.properties
-*.log
-
-/build
 .DS_STORE
+node_modules/
+build/
+.idea/
```
Comment (lipis) [inline .gitignore]: let's just sort them by name

File: example.js [hunk 7]
```diff
@@ -27,31 +24,27 @@ const ab = require("./build/Release/electron-addressbook")();
...
+console.log("Start Importing Contacts".custom);
 
+addressBook.getContacts(
+    progress => {
+        readline.clearLine(process.stdout, 0);
+        readline.cursorTo(process.stdout, 0);
+        process.stdout.write("Progress".custom + " " + progress + "%")
+    },
+    contacts => console.log("Contacts".custom, contacts)
+);
```
Comment (lipis) [inline example.js]: final end of line..

File: index.js [hunk 8]
```diff
@@ -15,4 +15,42 @@
...
+     *
+     * @param {Function} [onProgress] Callback provides overall process percent as an integer value between 1 to 100
+     * @param {Function} [onFinish] Callback provides an array contains all of the Addressbook contacts information
+     */
+    getContacts: function (onProgress, onFinish) {
+        return addressbook.getContacts(onProgress, onFinish)
+    }
+};
+
+module.exports = lib;
```
Comment (lipis) [inline index.js]: final end of line

File: wrapper.cc [hunk 9]
```diff
@@ -15,119 +15,133 @@
...
+
+    Nan::Set(target, New<String>("getContactsCount").ToLocalChecked(),
+        GetFunction(New<FunctionTemplate>(GetContactsCount)).ToLocalChecked());
 
-NODE_MODULE(addon, init)
+    Nan::Set(target, New<String>("getContacts").ToLocalChecked(),
+        GetFunction(New<FunctionTemplate>(GetContacts)).ToLocalChecked());
+}
 
+NODE_MODULE(addon, Init)
```
Comment (lipis) [inline wrapper.cc]: new line :)

**Comment by lipis** [comment]
@digital-flowers it's been a long wait.. but we are finally merged it! Thank you very much for your contribution :)

**Comment by digital-flowers** [comment]
Thank You Too for this Amazing Product :)

**Pull request merged**